src/java.desktop/share/classes/javax/swing/border/EtchedBorder.java

```diff
@@ -131,22 +151,61 @@ public EtchedBorder(int etchType, Color highlight, Color shadow)    {
      * @param height the height of the painted border
      */
     public void paintBorder(Component c, Graphics g, int x, int y, int width, int height) {
-        int w = width;
-        int h = height;
+        // We remove any initial transforms to prevent rounding errors
+        // when drawing in non-integer scales
+        AffineTransform at = new AffineTransform();
+        Stroke oldStk = new BasicStroke();
+        int stkWidth = 1;
+        boolean resetTransform = false;
+        if (g instanceof Graphics2D) {
+            Graphics2D g2d = (Graphics2D) g;
+            at = g2d.getTransform();
+            oldStk = g2d.getStroke();
+            // if m01 or m10 is non-zero, then there is a rotation or shear
+            // skip resetting the transform
+            resetTransform = (at.getShearX() == 0) && (at.getShearY() == 0);
+            if (resetTransform) {
+                g2d.setTransform(new AffineTransform());
+                stkWidth = (int) Math.floor(Math.min(at.getScaleX(), at.getScaleY()));
+                g2d.setStroke(new BasicStroke((float) stkWidth));
+            }
+            RenderingHints rend =
+                    new RenderingHints(RenderingHints.KEY_STROKE_CONTROL, RenderingHints.VALUE_STROKE_PURE);
+        }
 
-        g.translate(x, y);
+        int w;
+        int h;
+        int xtranslation;
+        int ytranslation;
+        if (resetTransform) {
+            w = (int) Math.floor(at.getScaleX() * width - 1);
+            h = (int) Math.floor(at.getScaleY() * height - 1);
+            xtranslation = (int) Math.ceil(at.getScaleX()*x+at.getTranslateX());
+            ytranslation = (int) Math.ceil(at.getScaleY()*y+at.getTranslateY());
+        } else {
+            w = width;
+            h = height;
+            xtranslation = x;
+            ytranslation = y;
+        }
 
-        g.setColor(etchType == LOWERED? getShadowColor(c) : getHighlightColor(c));
-        g.drawRect(0, 0, w-2, h-2);
+        g.translate(xtranslation, ytranslation);
 
-        g.setColor(etchType == LOWERED? getHighlightColor(c) : getShadowColor(c));
-        g.drawLine(1, h-3, 1, 1);
-        g.drawLine(1, 1, w-3, 1);
+        paintBorderShadow(g, (etchType == LOWERED) ? getHighlightColor(c)
+                                                   : getShadowColor(c),
+                          w, h, stkWidth);
+        paintBorderHighlight(g, (etchType == LOWERED) ? getShadowColor(c)
+                                                      : getHighlightColor(c),
+                             w, h, stkWidth);
 
-        g.drawLine(0, h-1, w-1, h-1);
-        g.drawLine(w-1, h-1, w-1, 0);
+        g.translate(-xtranslation, -ytranslation);
 
-        g.translate(-x, -y);
+        // Set the transform we removed earlier
+        if (g instanceof Graphics2D) {
```

Comment:
The condition should be
```suggestion
        if (resetTransform) {
```

If `resetTransform` is `true`, then `g` is an instance of `Graphics2D`.

If `resetTransform` is `false`, there's nothing to restore anyway.